File: Week_07/Java_Version/Nth_Dimensional/Vector.java
package Week_07.Java_Version.Nth_Dimensional;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import Week_07.Java_Version.Nth_Dimensional.Interface.Vector_Interface;

public class Vector implements Vector_Interface, Iterable<Double>, Cloneable {

    // Using a List to represent an n-dimensional vector.
    // This allows the vector to have any number of components (n),
    // unlike a fixed-size array which would be limited to 3 dimensions.
    private List<Double> vectorData;

    // A constructor for creating a vector of a specific dimension,
    // initialized with all components set to zero.
    public Vector(int dimension) {
        if (dimension < 0) {
            throw new IllegalArgumentException("Dimension cannot be negative.");
        }
        this.vectorData = new ArrayList<>();
        for (int i = 0; i < dimension; i++) {
            this.vectorData.add(0.0);
        }
    }

    // Constructor to initialize a vector from an array of doubles.
    // This provides a flexible way to create vectors of various dimensions.
    public Vector(double... data) {
        this.vectorData = new ArrayList<>();
        for (double value : data) {
            this.vectorData.add(value);
        }
    }

    // A copy constructor for creating a new Vector object
    // that is a deep copy of another Vector. This prevents
    // accidental modification of the original vector.
    public Vector(Vector other) {
        this.vectorData = new ArrayList<>(other.vectorData);
    }

    // Returns the number of dimensions of the vector.
    public int getDimension() {
        return vectorData.size();
    }

    @Override
    public double GET(int index) {
        if (index < 0 || index >= vectorData.size()) {
            throw new IndexOutOfBoundsException("Index " + index + " is out of bounds for vector of dimension " + getDimension() + ".");
        }
        return vectorData.get(index);
    }

    @Override
    public void SET(int index, double value) {
        if (index < 0 || index >= vectorData.size()) {
            throw new IndexOutOfBoundsException("Index " + index + " is out of bounds for vector of dimension " + getDimension() + ".");
        }
        vectorData.set(index, value);
    }

    @Override
    public double LENGTH() {
        double sumOfSquares = 0.0;
        // The length is calculated using the generalized Pythagorean theorem:
        // sum of the squares of all components.
        for (double component : vectorData) {
            sumOfSquares += component * component;
        }
        return Math.sqrt(sumOfSquares);
    }

    @Override
    public Vector ADD(Vector otherVector) {
        // We must ensure that both vectors have the same number of dimensions
        // to perform a valid addition.
        if (this.getDimension() != otherVector.getDimension()) {
            throw new IllegalArgumentException("Vectors must have the same dimension to be added.");
        }

        Vector result = new Vector(this.getDimension());
        for (int i = 0; i < this.getDimension(); i++) {
            result.SET(i, this.GET(i) + otherVector.GET(i));
        }
        return result;
    }

    @Override
    public Vector SUB(Vector otherVector) {
        // Similar to addition, subtraction also requires that both vectors
        // have the same number of dimensions.
        if (this.getDimension() != otherVector.getDimension()) {
            throw new IllegalArgumentException("Vectors must have the same dimension to be subtracted.");
        }

        Vector result = new Vector(this.getDimension());
        for (int i = 0; i < this.getDimension(); i++) {
            result.SET(i, this.GET(i) - otherVector.GET(i));
        }
        return result;
    }

    @Override
    public Vector CROSSPRODUCT(Vector otherVector) {
        // The cross product is only defined for 3-dimensional vectors.
        if (this.getDimension() != 3 || otherVector.getDimension() != 3) {
            throw new UnsupportedOperationException("Cross product is only defined for 3-dimensional vectors.");
        }

        double x1 = this.GET(0), y1 = this.GET(1), z1 = this.GET(2);
        double x2 = otherVector.GET(0), y2 = otherVector.GET(1), z2 = otherVector.GET(2);

        double newX = y1 * z2 - z1 * y2;
        double newY = z1 * x2 - x1 * z2;
        double newZ = x1 * y2 - y1 * x2;

        return new Vector(newX, newY, newZ);
    }

    @Override
    public double DOTPRODUCT(Vector otherVector) {
        if (this.getDimension() != otherVector.getDimension()) {
            throw new IllegalArgumentException("Vectors must have the same dimension for dot product.");
        }

        double dotProduct = 0.0;
        // The dot product is the sum of the products of corresponding components.
        // This loop works for any number of dimensions.
        for (int i = 0; i < this.getDimension(); i++) {
            dotProduct += this.GET(i) * otherVector.GET(i);
        }
        return dotProduct;
    }

    // A method to scale a vector by a scalar value.
    public Vector scale(double scalar) {
        Vector scaledVector = new Vector(this.getDimension());
        for (int i = 0; i < this.getDimension(); i++) {
            scaledVector.SET(i, this.GET(i) * scalar);
        }
        return scaledVector;
    }

    // This method returns a new vector that is the normalized version (unit vector)
    // of the original.
    public Vector normalize() {
        double length = this.LENGTH();
        if (length == 0) {
            // A zero vector cannot be normalized.
            throw new UnsupportedOperationException("Cannot normalize a zero vector.");
        }
        return this.scale(1.0 / length);
    }

    @Override
    public Iterator<Double> iterator() {
        return vectorData.iterator();
    }

    @Override
    public Object clone() throws CloneNotSupportedException {
        // The `super.clone()` performs a shallow copy.
        Vector cloned = (Vector) super.clone();
        // We then perform a deep copy of the `vectorData` list
        // to ensure the new vector is independent of the original.
        cloned.vectorData = new ArrayList<>(this.vectorData);
        return cloned;
    }

    @Override
    public String toString() {
        return "Vector(" + vectorData.toString().substring(1, vectorData.toString().length() - 1) + ")";
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null || this.getClass() != obj.getClass()) {
            return false;
        }
        Vector other = (Vector) obj;
        
        // Vectors of different dimensions cannot be equal.
        if (this.getDimension() != other.getDimension()) {
            return false;
        }

        // Using a tolerance (epsilon) is essential for comparing floating-point numbers.
        double epsilon = 1e-9;
        for (int i = 0; i < this.getDimension(); i++) {
            if (Math.abs(this.GET(i) - other.GET(i)) > epsilon) {
                return false;
            }
        }
        return true;
    }
}

File: Week_07/Python_Version/Class/Vector.py
import math
import copy

class Vector:
    """
    Represents a mathematical n-dimensional vector.

    This class provides methods for common vector operations such as addition,
    subtraction, dot product, cross product, magnitude, and normalization.
    """

    def __init__(self, *args) -> None:
        """
        Initializes a new Vector object.

        This single constructor handles multiple initialization patterns:
        - Vector(dimension): Creates a zero vector of the given dimension.
        - Vector(value1, value2, ...): Creates a vector from a variable number of arguments.
        - Vector([value1, value2, ...]): Creates a vector from a list or tuple.
        - Vector(other_vector): Creates a deep copy of another vector.
        """
        self.vector_data = []

        if len(args) == 0:
            # Default constructor: Creates a 0D vector.
            pass
        elif len(args) == 1:
            arg = args[0]
            if isinstance(arg, int):
                # Creates a zero vector of a specific dimension.
                if arg < 0:
                    raise ValueError("Dimension cannot be negative.")
                self.vector_data = [0.0] * arg
            elif isinstance(arg, (list, tuple)):
                # Creates a vector from a list or tuple.
                self.vector_data = [float(x) for x in arg]
            elif isinstance(arg, Vector):
                # Copy constructor: Creates a deep copy of another Vector.
                self.vector_data = copy.deepcopy(arg.vector_data)
            else:
                # Creates a vector of a single value.
                self.vector_data = [float(arg)]
        else:
            # Creates a vector from a variable number of arguments.
            self.vector_data = [float(x) for x in args]

    def get_dimension(self) -> int:
        """
        Returns the number of dimensions of the vector.
        """
        return len(self.vector_data)

    def get(self, index: int) -> float:
        """
        Gets the component at a specified index.
        """
        if index < 0 or index >= len(self.vector_data):
            raise IndexError(f"Index {index} is out of bounds for vector of dimension {self.get_dimension()}.")
        return self.vector_data[index]

    def set(self, index: int, value: float) -> None:
        """
        Sets the component at a specified index to a new value.
        """
        if index < 0 or index >= len(self.vector_data):
            raise IndexError(f"Index {index} is out of bounds for vector of dimension {self.get_dimension()}.")
        self.vector_data[index] = value

    def magnitude(self) -> float:
        """
        Calculates the magnitude (length) of the vector.
        """
        sum_of_squares = sum(x ** 2 for x in self.vector_data)
        return math.sqrt(sum_of_squares)

    def dot_product(self, other_vector: 'Vector') -> float:
        """
        Calculates the dot product of two vectors.
        """
        if self.get_dimension() != other_vector.get_dimension():
            raise ValueError("Vectors must have the same dimension for dot product.")

        dot_product = sum(a * b for a, b in zip(self.vector_data, other_vector.vector_data))
        return dot_product

    def cross_product(self, other_vector: 'Vector') -> 'Vector':
        """
        Calculates the cross product of two 3-dimensional vectors.
        """
        if self.get_dimension() != 3 or other_vector.get_dimension() != 3:
            raise TypeError("Cross product is only defined for 3-dimensional vectors.")

        x1, y1, z1 = self.vector_data
        x2, y2, z2 = other_vector.vector_data

        new_x = y1 * z2 - z1 * y2
        new_y = z1 * x2 - x1 * z2
        new_z = x1 * y2 - y1 * x2

        return Vector(new_x, new_y, new_z)

    def normalize(self) -> 'Vector':
        """
        Returns a new vector that is the normalized version (unit vector) of the original.
        """
        magnitude = self.magnitude()
        if magnitude == 0:
            raise ValueError("Cannot normalize a zero vector.")
        
        return self * (1.0 / magnitude)

    def copy(self) -> 'Vector':
        """
        Returns a deep copy of the vector.
        """
        return Vector(self)

    # Dunder methods for Python operator overloading and features

    def __str__(self) -> str:
        """
        Returns a string representation of the vector.
        Example: "Vector(1.0, 2.0, 3.0)"
        """
        return f"Vector({', '.join(str(x) for x in self.vector_data)})"

    def __repr__(self) -> str:
        """
        Returns a developer-friendly string representation.
        """
        return self.__str__()

    def __add__(self, other_vector: 'Vector') -> 'Vector':
        """
        Overloads the '+' operator for vector addition.
        """
        if self.get_dimension() != other_vector.get_dimension():
            raise ValueError("Vectors must have the same dimension for addition.")

        result_data = [a + b for a, b in zip(self.vector_data, other_vector.vector_data)]
        return Vector(result_data)

    def __sub__(self, other_vector: 'Vector') -> 'Vector':
        """
        Overloads the '-' operator for vector subtraction.
        """
        if self.get_dimension() != other_vector.get_dimension():
            raise ValueError("Vectors must have the same dimension for subtraction.")

        result_data = [a - b for a, b in zip(self.vector_data, other_vector.vector_data)]
        return Vector(result_data)

    def __mul__(self, scalar: float) -> 'Vector':
        """
        Overloads the '*' operator for scalar multiplication.
        """
        result_data = [x * scalar for x in self.vector_data]
        return Vector(result_data)

    def __rmul__(self, scalar: float) -> 'Vector':
        """
        Handles scalar multiplication when the scalar is on the left.
        Example: 5 * my_vector
        """
        return self.__mul__(scalar)

    def __truediv__(self, scalar: float) -> 'Vector':
        """
        Overloads the '/' operator for scalar division.
        """
        if scalar == 0:
            raise ZeroDivisionError("Cannot divide a vector by zero.")
        result_data = [x / scalar for x in self.vector_data]
        return Vector(result_data)

    def __getitem__(self, index: int) -> float:
        """
        Enables accessing vector components using bracket notation (e.g., my_vector[0]).
        """
        return self.get(index)

    def __setitem__(self, index: int, value: float) -> None:
        """
        Enables setting vector components using bracket notation (e.g., my_vector[0] = 5).
        """
        self.set(index, value)

    def __iter__(self):
        """
        Enables iterating over the vector components.
        """
        return iter(self.vector_data)

    def __eq__(self, other: object) -> bool:
        """
        Overloads the '==' operator for vector equality comparison.
        Compares vectors for equality using a tolerance for floating-point numbers.
        """
        if not isinstance(other, Vector):
            return NotImplemented

        if self.get_dimension() != other.get_dimension():
            return False

        for a, b in zip(self.vector_data, other.vector_data):
            if not math.isclose(a, b):
                return False
        return True

    def __len__(self) -> int:
        """
        Returns the number of dimensions.
        """
        return self.get_dimension()